sqlitecluster/SQLiteNode.cpp

```diff
@@ -1645,7 +1652,17 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
                 SDEBUG("Spawning concurrent replicate thread (blocks until DB handle available): " << threadID);
                 try {
                     uint64_t threadAttemptStartTimestamp = STimeNow();
-                    ResourceMonitorThread([=, this](){this->_replicate(peer, message, _dbPool->getIndex(false), threadAttemptStartTimestamp);}).detach();
+                    _replicateThreadStarted = false;
+                    thread(&SQLiteNode::_replicate, this, peer, message, _dbPool->getIndex(false), threadAttemptStartTimestamp).detach();
+                    {
+                        unique_lock<mutex> lock(_replicateStartMutex);
+                        while (!_replicateThreadStarted) {
+                            _replicateStartCV.wait(lock);
+                            if (!_replicateThreadStarted) {
+                                SINFO("condition variable finished waiting but replicate thread not started.");
```

Comment:
Just for my own education: is there any situation where this might be logged? Considering this will always wait on the notification on _replicateStartCV, and we're not using a timeout here, the only situation where we would leave this while loop is if the thread starts and _replicateThreadStarted is set to true, right?

Reply:
Also, could it happen that `_replicateStartCV.notify_all();` runs before we reach the line `_replicateStartCV.wait(lock);` above?

Reply:
It's possible because of this: https://en.wikipedia.org/wiki/Spurious_wakeup

See the `condition_variable` documentation here:
https://en.cppreference.com/w/cpp/thread/condition_variable

Specifically:
> Call [wait](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for), or [wait_until](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until) on the std::condition_variable (atomically releases the mutex and suspends thread execution until the condition variable is notified, a timeout expires, or a [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) occurs, then atomically acquires the mutex before returning).

> Also, could it happen that _replicateStartCV.notify_all(); runs before we reach the line _replicateStartCV.wait(lock); above?

Yes, but in that case we will never get to that line, `_replicateThreadStarted` will already be true when we get to the `while` loop.

See the notes here for some detail:
https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

You get to read the only use I've ever come across of the word "pessimization".

Reply:
I see, thanks.

 The reason why you didn't use a predicate by overloading the wait method here is just so you could have that SINFO, or is it preferable to write that while loop on the condition?